library/src/main/java/com/heinrichreimersoftware/materialintro/app/IntroActivity.java

```diff
@@ -359,6 +359,28 @@ else if (position < currentPosition && Math.ceil(position) != currentPosition &&
     private long calculateScrollDuration(int distance) {
         return Math.round(pageScrollDuration * (distance + Math.sqrt(distance)) / 2);
     }
+    
+    public void goToFirstSlide() {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem > adapter.getCount() - 1) finishIfNeeded();
+
+        if (canGoForward(currentItem, true)) {
+            smoothScrollPagerTo(0);
+        } else {
+            AnimUtils.applyShakeAnimation(this, buttonNext);
+        }
+    }
+
+    public void goToLastSlide() {
+        int currentItem = pager.getCurrentItem();
+        if (currentItem > adapter.getCount() - 1) finishIfNeeded();
+
+        if (canGoForward(currentItem, true)) {
```

Comment:
Here it is necessary to check `canGoForward()` for *all* slides between the current and the last slide.